lib/truffle/truffle/patching.rb

```diff
@@ -1,140 +1,68 @@
-There is also a potential patch available '#{patch_candidate}' for the file.
-No patches are applied. Please add the gem's lib to $LOAD_PATH first to fix.
-      TXT
-    end
+    original = originals[name.to_s].flat_map do |original_path|
+      Pathname.glob(File.join(original_path, require_path))
+    end.first
 
-    nil
+    require original
   end
 
-  def self.required?(relative_path)
-    $LOADED_FEATURES.include? find_absolute_path(relative_path).to_s
-  end
-end
-
-module Kernel
+  def install_gem_activation_hook
+    Gem::Specification.class_eval do
+      alias_method :activate_without_truffle_patching, :activate
 
-  private
-
-  alias_method :require_without_truffle_patching, :gem_original_require
-
-  def gem_original_require(path)
-    path           = Pathname(Rubinius::Type.coerce_to_path(path))
-    path_no_suffix = path.sub_ext ""
-    relative_path  = if path_no_suffix.absolute?
-                       Truffle::Patching.find_relative_path path_no_suffix.sub_ext('.rb')
-                     else
-                       path_no_suffix
-                     end
-
-    operation = Truffle::Patching::PATCHES[relative_path.to_s]
-
-    # not patched requiring normally
-    return require_without_truffle_patching path if operation.nil?
-
-    # apply patch based on operation
-    log                 = -> { Truffle::System.log :PATCH, "#{operation} original require '#{path}'" }
-    absolute_patch_path = Truffle::Patching::DIR.join(path_no_suffix).sub_ext(".rb")
-    case operation
-    when :after
-      require_without_truffle_patching(path.to_s).tap do |required|
-        if required
-          log.call
-          require_without_truffle_patching absolute_patch_path.to_s
-        end
-      end
-    when :before
-      if Truffle::Patching.required? relative_path
-        result = require_without_truffle_patching path.to_s
-        raise 'should be already loaded' if result
+      def activate
+        result = activate_without_truffle_patching
+        Truffle::Patching.insert_patching_dir name, *full_require_paths
         result
-      else
-        log.call
-        require_without_truffle_patching absolute_patch_path.to_s
-        require_without_truffle_patching path.to_s
       end
-    when :instead
-      log.call
-      require_without_truffle_patching absolute_patch_path.to_s
-    when :ignored
-      require_without_truffle_patching path.to_s
-    else
-      raise "unrecognized operation #{operation} for #{relative_path}"
     end
   end
-end
 
-class Module
+  private
 
-  alias_method :autoload_without_truffle_patching, :autoload
-  private :autoload_without_truffle_patching
+  def originals
+    @originals ||= {}
+  end
 
-  def autoload(const, path)
-    if Truffle::Patching::PATCHES[path]
-      require path
-    else
-      autoload_without_truffle_patching const, path
-    end
+  def dir
+    @dir ||= Pathname(Truffle::Boot.ruby_home).join('lib/patches')
   end
+
+  extend self
```

Comment:
In the code I've seen it was usually at the bottom, but makes sense to put it in a usual place of includes and extends.